api/api.proto

```diff
@@ -63,6 +63,149 @@ message Pump {
+
+  // guard rails
+  GuardRails guardRails = 5;
+}
+
+message GuardRails {
+  // Suspend threshold guard rail
+  SuspendThresholdGuardRail suspendThreshold = 1;
+
+  // Insulin sensitivity guard rail
+  InsulinSensitivityGuardRail insulinSensitivity = 2;
+
+  // Basal rates guard rail
+  BasalRatesGuardRail basalRates = 3;
+
+  // Carbohydrate ratio guard rail
+  CarbohydrateRatioGuardRail carbohydrateRatio = 4;
+
+  // Basal rate maximum guard rail
+  BasalRateMaximumGuardRail basalRateMaximum = 5;
+
+  // Bolus amount maximum guard rail
+  BolusAmountMaximumGuardRail bolusAmountMaximum = 6;
+
+  // Correction range guard rail
+  CorrectionRangeGuardRail correctionRange = 7;
+}
+
+message SuspendThresholdGuardRail {
+  // Blood glucose units
+  BloodGlucoseUnits units = 1;
+
+  // Tidepool recommended bounds
+  RecommendedBounds recommendedBounds = 2;
+
+  // Device specific absolute bounds
+  AbsoluteBounds absoluteBounds = 3;
+}
+
+message InsulinSensitivityGuardRail {
+  // Blood glucose units
+  BloodGlucoseUnits units = 1;
+
+  // Tidepool recommended bounds
+  RecommendedBounds recommendedBounds = 2;
+
+  // Device specific absolute bounds
+  AbsoluteBounds absoluteBounds = 3;
+}
+
+message BasalRatesGuardRail {
+  // Basal rate units
+  BasalRateUnits units = 1;
+
+  // Default value
+  double defaultValue = 2;
+
+  // Device specific absolute bounds. Some pumps might have different increments for different ranges.
+  repeated AbsoluteBounds absoluteBounds = 3;
+}
+
+message CarbohydrateRatioGuardRail {
+  // Carbohydrate ratio units
+  CarbohydrateRatioUnits units = 1;
+
+  // Tidepool recommended bounds
+  RecommendedBounds recommendedBounds = 2;
+
+  // Device specific absolute bounds
+  AbsoluteBounds absoluteBounds = 3;
+}
+
+message BasalRateMaximumGuardRail {
+  // Basal rate units
+  BasalRateUnits units = 1;
+
+  // Default value
+  double defaultValue = 2;
+
+  // Device specific absolute bounds
+  AbsoluteBounds absoluteBounds = 3;
+}
+
+message BolusAmountMaximumGuardRail {
+  // Bolus units
+  BolusUnits units = 1;
+
+  // Default value
+  double defaultValue = 2;
+
+  // Tidepool recommended bounds
+  RecommendedBounds recommendedBounds = 3;
+
+  // Device specific absolute bounds
+  AbsoluteBounds absoluteBounds = 4;
+}
+
+message CorrectionRangeGuardRail {
+  // Blood glucose units
+  BloodGlucoseUnits units = 1;
+
+  // Tidepool recommended bounds
+  RecommendedBounds recommendedBounds = 2;
+
+  // Device specific absolute bounds
+  AbsoluteBounds absoluteBounds = 3;
+}
+
+// Closed range double interval
+message AbsoluteBounds {
+  // Lower bound (inclusive)
+  double minimum = 1;
+
+  // Upper bound (inclusive)
+  double maximum = 2;
+
+  // Increment
+  double increment = 3;
+}
+
+// Closed range double interval
+message RecommendedBounds {
+  // Lower bound (inclusive)
+  double minimum = 1;
+
+  // Upper bound (inclusive)
+  double maximum = 2;
+}
+
+enum BloodGlucoseUnits {
+  MilligramsPerDeciliter = 0;
```

Comment:
When this is marshaled to json it is represented as a string with value `MilligramsPerDeciliter` - so far I haven't found a way to customize it (so it matches the units in the data service - e.g. `mg/dl`). Even if there is way for this to be done, it break the spec as proto-to-json mapping is defined [here](https://developers.google.com/protocol-buffers/docs/proto3#json).

Reply:
For internal APIs, I don't think this is a problem. But for anything external, those representations `mg/dL` and `mmol/L` are pretty universal (see [Open mHealth's Blood Glucose type](https://www.openmhealth.org/schemas/omh_blood-glucose/)). Can we use Enum Value Options as described [here](https://stackoverflow.com/a/19258863) or [here](https://developers.google.com/protocol-buffers/docs/proto#customoptions)?

Reply:
I saw those threads on SO too. Custom options are a way for adding metadata to proto types. They must be interpreted by something. We would then have to write our own serializer and figure out how to make it work with grpc-gateway. This functionality might be supported by `gogo/protobuf`, which is a fork of protobuf, but the project future seems uncertain. Another option is to use strings, but then each service will have to check the string value and will not be as type safe as using an enum.